components/lang/src/main/java/io/cellery/impl/CreateInstance.java

```diff
@@ -81,32 +124,144 @@
+        String instanceArg;
+        boolean startDependencies = ctx.getBooleanArgument(0);
+        shareDependencies = ctx.getBooleanArgument(1);
         final BMap refArgument = (BMap) ctx.getNullableRefArgument(0);
         LinkedHashMap nameStruct = ((BMap) ctx.getNullableRefArgument(1)).getMap();
         String cellName = ((BString) nameStruct.get("name")).stringValue();
-        String instanceName = ((BString) nameStruct.get(INSTANCE_NAME)).stringValue();
-        String destinationPath = System.getenv(CELLERY_IMAGE_DIR_ENV_VAR) + File.separator +
+        isRoot = ((BBoolean) nameStruct.get("isRoot")).booleanValue();
+        instanceName = ((BString) nameStruct.get(INSTANCE_NAME)).stringValue();
+        if (instanceName.isEmpty()) {
+            instanceName = generateRandomInstanceName(((BString) nameStruct.get("name")).stringValue(),
+                    ((BString) nameStruct.get("ver")).stringValue());
+        }
+        String cellImageDir = System.getenv(CELLERY_IMAGE_DIR_ENV_VAR);
+        if (cellImageDir == null) {
+            try (InputStream inputStream = new FileInputStream(DEBUG_BALLERINA_CONF)) {
+                Properties properties = new Properties();
+                properties.load(inputStream);
+                cellImageDir = properties.getProperty(CELLERY_IMAGE_DIR_ENV_VAR).replaceAll("\"", "");
+            } catch (IOException e) {
+                throw new BallerinaException("Unable to read " + DEBUG_BALLERINA_CONF, e);
+            }
+        }
+        String destinationPath = cellImageDir + File.separator +
                 "artifacts" + File.separator + "cellery";
-        Map dependencyInfo = ((BMap) ctx.getNullableRefArgument(2)).getMap();
+        Map userDependencyLinks = ((BMap) ctx.getNullableRefArgument(2)).getMap();
+
         String cellYAMLPath = destinationPath + File.separator + cellName + YAML;
         Composite composite;
         if (CELL.equals(((BString) (refArgument.getMap().get("kind"))).stringValue())) {
+            instanceArg = "cells.mesh.cellery.io/" + instanceName;
             composite = CelleryUtils.readCellYaml(cellYAMLPath);
         } else {
+            instanceArg = "composites.mesh.cellery.io/" + instanceName;
             composite = CelleryUtils.readCompositeYaml(cellYAMLPath);
         }
+        if (isRoot) {
+            // Must generate dependency tree regardless of whether starting dependencies or not
+            // if the instance is root
+            try {
+                generateDependencyTree(destinationPath + File.separator + "metadata.json");
+                if (!startDependencies) {
+                    if (((Meta) dependencyTree.getRoot().getData()).getDependencies().size() > 0) {
+                        validateRootDependencyLinks(userDependencyLinks);
+                    }
+                }
+            } catch (IOException e) {
+                String error = "Unable to generate dependency tree";
+                log.error(error, e);
```

Comment:
shouldn't this error be propagated up and handled?

Reply:
Fixed in all places.